dartagnan/src/main/java/com/dat3m/dartagnan/witness/graphviz/ExecutionGraphVisualizer.java

```diff
@@ -33,151 +32,140 @@ public class ExecutionGraphVisualizer {
-                .forEach(entry -> sortedMemoryObjects.add(entry.getValue()));
+             .sorted(Comparator.comparing(entry -> entry.getValue().address()))
+             .forEach(entry -> sortedMemoryObjects.add(entry.getValue()));
     }
 
-    private boolean ignore(EventData e) {
+    private boolean ignore(EventModel e) {
         return false; // We ignore no events for now.
     }
 
-    private ExecutionGraphVisualizer addReadFrom(ExecutionModel model) {
-
-        graphviz.beginSubgraph("ReadFrom");
-        graphviz.setEdgeAttributes("color=green");
-        for (Map.Entry<EventData, EventData> rw : model.getReadWriteMap().entrySet()) {
-            EventData r = rw.getKey();
-            EventData w = rw.getValue();
-
-            if (ignore(r) || ignore(w) || !rfFilter.test(w, r)) {
-                continue;
+    private ExecutionGraphVisualizer addRelations(ExecutionModelNext model) {
+        for (RelationModel rm : model.getRelationModels()) {
+            if (rm.hasName(PO)) {
+                addProgramOrder(model);
+            } else {
+                addRelation(model, rm);
             }
-
-            appendEdge(w, r, "label=rf");
         }
-        graphviz.end();
         return this;
     }
 
-    private ExecutionGraphVisualizer addFromRead(ExecutionModel model) {
-
-        graphviz.beginSubgraph("FromRead");
-        graphviz.setEdgeAttributes("color=orange");
-        for (Map.Entry<EventData, EventData> rw : model.getReadWriteMap().entrySet()) {
-            EventData r = rw.getKey();
-            EventData w = rw.getValue();
-
-            if (ignore(r) || ignore(w)) {
-                continue;
-            }
-
-            List<EventData> co = model.getCoherenceMap().get(w.getAccessedAddress());
-            // Check if exists w2 : co(w, w2)
-            if (co.indexOf(w) + 1 < co.size()) {
-                EventData w2 = co.get(co.indexOf(w) + 1);
-                if (!ignore(w2) && frFilter.test(r, w2)) {
-                    appendEdge(r, w2, "label=fr");
-                }
-            }
+    private ExecutionGraphVisualizer addRelation(ExecutionModelNext model, RelationModel rm) {
+        String name = rm.getName();
+        graphviz.beginSubgraph(name);
+        String attributes = String.format("color=%s", colorMap.getColor(name));
+        graphviz.setEdgeAttributes(attributes);
+        String label = String.format("label=\"%s\"", name);
+        BiPredicate<EventModel, EventModel> filter;
+        if (name.equals(CO)) {
+            filter = coFilter;
+        } else if (name.equals(RF)) {
+            filter = rfFilter;
+        } else if (name.equals("fr")) {
+            filter = frFilter;
+        } else { filter = getFilter(name); }
+        for (RelationModel.EdgeModel edge : rm.getEdgeModels()) {
+            EventModel from = edge.getFrom();
+            EventModel to = edge.getTo();
+
+            if (ignore(from) || ignore(to) || !filter.test(from, to)) { continue; }
```

Comment:
I think the `ignore` methods are obsolete now and you should get rid of them

Reply:
Maybe. But I think they can make sense if `ExecutionModel(Next)`'s only purpose is to represent the execution itself but not how it should be printed/displayed.